File: src/main/java/com/koddy/server/member/domain/repository/query/spec/SearchMenteeCondition.java
package com.koddy.server.member.domain.repository.query.spec;

import com.koddy.server.member.domain.model.Language;
import com.koddy.server.member.domain.model.Nationality;
import org.springframework.util.CollectionUtils;

import java.util.Collections;
import java.util.List;

public record SearchMenteeCondition(
        NationalityCondition nationality,
        LanguageCondition language
) {
    public record NationalityCondition(
            boolean contains,
            List<Nationality> values
    ) {
    }

    public record LanguageCondition(
            boolean contains,
            List<Language.Category> values
    ) {
    }

    public static SearchMenteeCondition basic() {
        return new SearchMenteeCondition(
                createNationalityCondition(Collections.emptyList()),
                createLanguageCondition(Collections.emptyList())
        );
    }

    public static SearchMenteeCondition of(
            final List<Nationality> nationalities,
            final List<Language.Category> languages
    ) {
        return new SearchMenteeCondition(
                createNationalityCondition(nationalities),
                createLanguageCondition(languages)
        );
    }

    private static NationalityCondition createNationalityCondition(final List<Nationality> nationalities) {
        if (CollectionUtils.isEmpty(nationalities)) {
            return new NationalityCondition(false, Collections.emptyList());
        }
        return new NationalityCondition(true, nationalities);
    }

    private static LanguageCondition createLanguageCondition(final List<Language.Category> languages) {
        if (CollectionUtils.isEmpty(languages)) {
            return new LanguageCondition(false, Collections.emptyList());
        }
        return new LanguageCondition(true, languages);
    }
}

File: src/main/java/com/koddy/server/member/domain/repository/query/spec/SearchMenteeCondition.kt
package com.koddy.server.member.domain.repository.query.spec

import com.koddy.server.member.domain.model.Language
import com.koddy.server.member.domain.model.Nationality

data class SearchMenteeCondition(
    val nationality: NationalityCondition,
    val language: LanguageCondition,
) {
    data class NationalityCondition(
        val exists: Boolean,
        val values: List<Nationality>,
    )

    data class LanguageCondition(
        val exists: Boolean,
        val values: List<Language.Category>,
    )

    companion object {
        fun basic(): SearchMenteeCondition {
            return SearchMenteeCondition(
                nationality = NationalityCondition(exists = false, values = emptyList()),
                language = LanguageCondition(exists = false, values = emptyList()),
            )
        }

        fun of(
            nationalities: List<Nationality>,
            languages: List<Language.Category>,
        ): SearchMenteeCondition {
            return SearchMenteeCondition(
                nationality = createNationalityCondition(nationalities),
                language = createLanguageCondition(languages),
            )
        }

        private fun createNationalityCondition(nationalities: List<Nationality>): NationalityCondition {
            return when {
                nationalities.isEmpty() -> NationalityCondition(exists = false, values = emptyList())
                else -> NationalityCondition(exists = true, values = nationalities)
            }
        }

        private fun createLanguageCondition(languages: List<Language.Category>): LanguageCondition {
            return when {
                languages.isEmpty() -> LanguageCondition(exists = false, values = emptyList())
                else -> LanguageCondition(exists = true, values = languages)
            }
        }
    }
}
